### PR TITLE
refactor slurm_job_id

### LINKED ISSUE #10615: Deprecate `trainer.slurm_job_id`
## Proposed refactor

Deprecate the trainer property here: https://github.com/PyTorchLightning/pytorch-lightning/blob/6b728713bb3b35ad58cd0085acaa443b33ab03ac/pytorch_lightning/trainer/trainer.py#L1731-L1744

### Motivation

The property can be a static utility function instead. Doing this will consolidate SLURM environment variable code in Lightning and simplify the Trainer.

### Pitch

1. Move this logic to the `SLURMEnvironment` class as a staticmethod 
```python:
@staticmethod
def slurm_job_id() -> Optional[int]:
```
2. Then replace usages of `trainer.slurm_job_id` with `SLURMEnvironment.slurm_job_id()`
3. Then deprecate `trainer.slurm_job_id` 

### Additional context

______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

- [**Metrics**](https://github.com/PyTorchLightning/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Lite**](https://pytorch-lightning.readthedocs.io/en/latest/starter/lightning_lite.html): enables pure PyTorch users to scale their existing code on any kind of device while retaining full control over their own loops and optimization logic.

- [**Flash**](https://github.com/PyTorchLightning/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, fine-tuning, and solving problems with deep learning.

- [**Bolts**](https://github.com/PyTorchLightning/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks, and more for research and production with PyTorch Lightning and PyTorch.

- [**Lightning Transformers**](https://github.com/PyTorchLightning/lightning-transformers): Flexible interface for high-performance research using SOTA Transformers leveraging Pytorch Lightning, Transformers, and Hydra.

### PR DESCRIPTION
**Pull request by AndresAlgaba**
## What does this PR do?
Fixes #10615 

### Does your PR introduce any breaking changes? If yes, please list them.
None

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/plugins/environments/slurm_environment.py [hunk 1]
```diff
@@ -42,6 +43,21 @@ def detect() -> bool:
         """Returns ``True`` if the current process was launched on a SLURM cluster."""
         return "SLURM_NTASKS" in os.environ
 
+    @staticmethod
+    def job_id() -> Optional[int]:
```
Comment (awaelchli) [inline pytorch_lightning/plugins/environments/slurm_environment.py:47]: we can drastically improve code quality here in a follow up PR

**Review by awaelchli — CHANGES_REQUESTED** [review]
Good start! thanks for working on it :)

File: pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py [hunk 2]
```diff
@@ -81,7 +83,7 @@ def configure_logger(self, logger: Union[bool, LightningLoggerBase, Iterable[Lig
             # default logger
             self.trainer.logger = (
                 TensorBoardLogger(
-                    save_dir=self.trainer.default_root_dir, version=self.trainer.slurm_job_id, name="lightning_logs"
+                    save_dir=self.trainer.default_root_dir, version=self.env.job_id, name="lightning_logs"
```
[the remark this answers is not in the document]
Comment (AndresAlgaba) [reply pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py:86]: Thanks for the feedback! Yes, I saw this, should have noticed it already in the local mypy testing, sorry!

File: pytorch_lightning/plugins/environments/slurm_environment.py (hunk 1 above)
Comment (ananthsub) [reply pytorch_lightning/plugins/environments/slurm_environment.py:47]: @awaelchli would you prefer introducing `job_name` as a separate staticmethod in this PR itself?

File: pytorch_lightning/trainer/trainer.py [hunk 3]
```diff
@@ -1728,21 +1728,6 @@ def use_amp(self) -> bool:
     def is_global_zero(self) -> bool:
         return self.global_rank == 0
 
-    @property
-    def slurm_job_id(self) -> Optional[int]:
```
[the remark this answers is not in the document]
Comment (AndresAlgaba) [reply pytorch_lightning/trainer/trainer.py:1732]: I will follow up on this on Monday if that's okay? Already thanks for your feedback!

[the remark this answers is not in the document]
Comment (AndresAlgaba) [reply pytorch_lightning/trainer/trainer.py:1732]: I've added the slum_job_id again to the Trainer with a deprecation warning, and added a test for it!

File: pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (AndresAlgaba) [reply pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py]: I've added it now statically!

File: pytorch_lightning/plugins/environments/slurm_environment.py (hunk 1 above)
Comment (AndresAlgaba) [reply pytorch_lightning/plugins/environments/slurm_environment.py:42]: Should it be included in this PR, or just the deprecation of job_id in Trainer?

File: pytorch_lightning/trainer/trainer.py [hunk 4]
```diff
@@ -1730,17 +1731,8 @@ def is_global_zero(self) -> bool:
...
-            except ValueError:
-                job_id = None
-
-        # in interactive mode, don't make logs use the same job id
-        in_slurm_interactive_mode = os.environ.get("SLURM_JOB_NAME") == "bash"
-        if in_slurm_interactive_mode:
-            job_id = None
+        rank_zero_deprecation("Method `slurm_job_id` is deprecated in v1.6.0 and will be removed in v1.7.0.")
+        job_id = SLURMEnvironment.job_id()
         return job_id
```
Comment (tchaton) [inline pytorch_lightning/trainer/trainer.py:1736]:
```suggestion
        return SLURMEnvironment.job_id()
```

**Review by tchaton — APPROVED** [review]
LGTM !

**Review by carmocca — APPROVED** [review]

File: pytorch_lightning/trainer/trainer.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (awaelchli) [reply pytorch_lightning/trainer/trainer.py:1732]: looks great!

File: pytorch_lightning/plugins/environments/slurm_environment.py (hunk 1 above)
Comment (awaelchli) [reply pytorch_lightning/plugins/environments/slurm_environment.py:42]: just the deprecation here, no further changes :) 
@ananthsub yes that sounds like a good idea to me! @AndresAlgaba interested for doing this in a follow-up PR?

**Review by awaelchli — APPROVED** [review]

File: CHANGELOG.md [hunk 5]
```diff
@@ -62,7 +62,7 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Deprecated the `precision_plugin` constructor argument from `Accelerator` ([#10570](https://github.com/PyTorchLightning/pytorch-lightning/pull/10570))
 
 
--
+- Deprecated the property `Trainer.slurm_job_id` in favor of the new `SLURMEnvironment.slurm_job_id()` method ([#10622](https://github.com/PyTorchLightning/pytorch-lightning/pull/10622))
```
Comment (ananthsub) [inline CHANGELOG.md:65]:
```suggestion
- Deprecated the property `Trainer.slurm_job_id` in favor of the new `SLURMEnvironment.job_id()` method ([#10622](https://github.com/PyTorchLightning/pytorch-lightning/pull/10622))
```

**Review by ananthsub — APPROVED** [review]

File: pytorch_lightning/plugins/environments/slurm_environment.py (hunk 1 above)
Comment (AndresAlgaba) [reply pytorch_lightning/plugins/environments/slurm_environment.py:42]: Yes, sure!
